### PR TITLE
Guggy Integration

### PR DESCRIPTION
**Pull request by lukebarnard1**
Very much like Giphy, except Guggy will implant the query you pass it into a generated GIF.

### CODE REVIEW

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/go-neb/matrix"
+	"github.com/matrix-org/go-neb/plugin"
+	"github.com/matrix-org/go-neb/types"
+	"net/http"
+	"strconv"
+	"strings"
+)
+type guggyQuery struct {
+	// "mp4" of "gif"
```
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: `or`

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	ReqID string `json:"reqId"`
+	GIF string `json:"gif"`
+	Width float64 `json:"width"`
+	Height float64 `json:"height"`
+}
+
+type guggyService struct {
+	id            string
+	serviceUserID string
+	APIKey        string
```
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: `api_key` please

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	}
+}
+func (s *guggyService) cmdGuggy(client *matrix.Client, roomID, userID string, args []string) (interface{}, error) {
+	// only 1 arg which is the text to search for.
+	querySentence := strings.Join(args, " ")
+	gifResult, err := s.text2gifGuggy(querySentence)
+	if err != nil {
+		return nil, err
+	}
+	mxc, err := client.UploadLink(gifResult.GIF)
```
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: As we found when debugging an earlier problem, if `gifResult.GIF == ""` then this fails with an obscure error message. Can we please check for the empty string and fail more gracefully.

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	}
+	var result guggyGifResult
+	if err := json.NewDecoder(res.Body).Decode(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func asInt(strInt string) uint {
```
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: Unused function. Please remove.

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go (hunk 2 above)
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: If anything, should it not be ```apiKey```? (giphy.go has the same problem)

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	query.Sentence = querySentence
+
+	reqBody, err := json.Marshal(query);
+	if err != nil {
+		return nil, err
+	}
+
+	reader := bytes.NewReader(reqBody)
+
+	req, err := http.NewRequest("POST", "http://text2gif.guggy.com/guggify", reader)
```
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: No HTTPS support?

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go (hunk 2 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: No, it has to be public in order for `json.Marshal` to see it. Over the wire we should be consistently using `keys_with_underscores`.

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go (hunk 5 above)
Comment (lukebarnard1) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: Yes actually!

File: src/github.com/matrix-org/go-neb/services/guggy/guggy.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	ReqID string `json:"reqId"`
+	GIF string `json:"gif"`
+	Width float64 `json:"width"`
+	Height float64 `json:"height"`
+}
+
+type guggyService struct {
+	id            string
+	serviceUserID string
+	api_key       string
```
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/guggy/guggy.go]: This needs to be:
```go
APIKey string `json:"api_key"`
```

**Comment by Kegsay** [comment]
LGTM other than `APIKey`.

**Pull request merged**